File: Algo/Leetcode/239MaxValInSlidingWindow.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        deque<int> window;
        vector<int> res;
        if(nums.size()==0 || k ==0) return res;
        for(int i=0; i<nums.size(); i++){
            while(!window.empty() && window.front() <= i-k)
                window.pop_front();
            while(!window.empty() && nums[window.back()] < nums[i])
                window.pop_back();
            window.push_back(i);
            if(i>=k-1) res.push_back(nums[window.front()]);
        }

        return res;
    }
};

File: Algo/Leetcode/239MaxValInSlidingWindow.py
class Solution(object):
    def maxSlidingWindow(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        q = []
        res = []
        for i in range(len(nums)):
            while len(q) > 0 and q[0] <= i-k:    # too old
                q.pop(0)
            while len(q) > 0 and nums[q[-1]] < nums[i]:   # most cur but small ones
                q.pop()
            q.append(i)
            if i >= k-1:
                res.append(nums[q[0]])

        return res
